test/e2e/shared/common.go

```diff
@@ -262,3 +265,122 @@ func SetEnvVar(key, value string, private bool) {
 	Logf("Setting environment variable: key=%s, value=%s", key, printableValue)
 	_ = os.Setenv(key, value)
 }
+
+// getOpenStackClusterFromMachine gets the OpenStackCluster that is related to the given machine.
+func getOpenStackClusterFromMachine(ctx context.Context, client client.Client, machine *clusterv1.Machine) (*infrav1.OpenStackCluster, error) {
+	key := types.NamespacedName{
+		Namespace: machine.Namespace,
+		Name:      machine.Spec.ClusterName,
+	}
+	cluster := &clusterv1.Cluster{}
+	err := client.Get(ctx, key, cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	key = types.NamespacedName{
+		Namespace: cluster.Spec.InfrastructureRef.Namespace,
+		Name:      cluster.Spec.InfrastructureRef.Name,
+	}
+	openStackCluster := &infrav1.OpenStackCluster{}
+	err = client.Get(ctx, key, openStackCluster)
+	return openStackCluster, err
+}
+
+type OpenStackLogCollector struct {
+	E2EContext E2EContext
+}
+
+// CollectMachineLog gets logs for the OpenStack resources related to the given machine.
+func (o OpenStackLogCollector) CollectMachineLog(ctx context.Context, managementClusterClient client.Client, m *clusterv1.Machine, outputPath string) error {
+	machineLogBase := path.Join(outputPath, "instances", m.Namespace, m.Name)
+	metaLog := path.Join(machineLogBase, "instance.log")
+
+	if err := os.MkdirAll(filepath.Dir(metaLog), 0o750); err != nil {
+		_, _ = fmt.Fprintf(GinkgoWriter, "couldn't create directory %q for file: %s\n", metaLog, err)
+	}
+
+	f, err := os.OpenFile(metaLog, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
+	if err != nil {
+		_, _ = fmt.Fprintf(GinkgoWriter, "couldn't open file %q: %s\n", metaLog, err)
+		return nil
+	}
+	defer f.Close()
+
+	openStackCluster, err := getOpenStackClusterFromMachine(ctx, managementClusterClient, m)
+	if err != nil {
+		return fmt.Errorf("error getting OpenStackCluster for Machine: %s", err)
+	}
+
+	ip := m.Status.Addresses[0].Address
+
+	srvs, err := GetOpenStackServers(&o.E2EContext, openStackCluster, sets.New(m.Spec.InfrastructureRef.Name))
+	if err != nil {
+		return fmt.Errorf("cannot dump machines, could not get servers from OpenStack: %v", err)
+	}
+	if len(srvs) != 1 {
+		return fmt.Errorf("expected exactly 1 server but got %d", len(srvs))
+	}
+	srv := srvs[m.Spec.InfrastructureRef.Name]
+
+	serverJSON, err := json.MarshalIndent(srv, "", "    ")
+	if err != nil {
+		return fmt.Errorf("error marshalling server %v: %s", srv, err)
+	}
+	if err := os.WriteFile(path.Join(machineLogBase, "server.txt"), serverJSON, 0o600); err != nil {
+		return fmt.Errorf("error writing server JSON %s: %s", serverJSON, err)
+	}
+	_, _ = fmt.Fprintf(f, "instance found: %q\n", srv.ID)
+
+	srvUser := o.E2EContext.E2EConfig.GetVariable(SSHUserMachine)
+	executeCommands(
+		ctx,
+		o.E2EContext.Settings.ArtifactFolder,
+		o.E2EContext.Settings.Debug,
+		filepath.Dir(f.Name()),
```

Comment:
I must confess I just copied this from [dumpMachine](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/bc57b7f750977aebf17ec820cfa5030dc72281f2/test/e2e/shared/common.go#L186). We do write something to that file, although I do not see how it is useful. It is this one: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-openstack/1581/pull-cluster-api-provider-openstack-e2e-test/1668516451239071744/artifacts/clusters/bootstrap/instances/e2e-89o9gs/cluster-e2e-89o9gs-control-plane-dl7hx/instance.log

> instance found: "fe61d653-7997-483c-ac73-839cae7fe3ce"

The content comes from this line in `dumpMachine` or line 333 here above: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/bc57b7f750977aebf17ec820cfa5030dc72281f2/test/e2e/shared/common.go#L181

`executeCommands` will not write anything to the file. It will instead write a new file in the folder for each of the commands we send.